src/dispatcher.rs

```diff
@@ -427,6 +427,27 @@ impl Dispatcher {
         }
     }
 
+    fn on_foreign_function(
+        &self,
+        context_id: u32,
+        function_id: u32,
+        arugments_size: usize,
+    ) {
+        if let Some(http_stream) = self.http_streams.borrow_mut().get_mut(&context_id) {
+            self.active_id.set(context_id);
+            hostcalls::set_effective_context(context_id).unwrap();
+            http_stream.on_foreign_function(function_id, arugments_size)
+        } else if let Some(stream) = self.streams.borrow_mut().get_mut(&context_id) {
+            self.active_id.set(context_id);
+            hostcalls::set_effective_context(context_id).unwrap();
+            stream.on_foreign_function(function_id, arugments_size)
+        } else if let Some(root) = self.roots.borrow_mut().get_mut(&context_id) {
+            self.active_id.set(context_id);
+            hostcalls::set_effective_context(context_id).unwrap();
+            root.on_foreign_function(function_id, arugments_size)
+        }
+    }
+
```

Comment:
Nit: Please move this towards end of this block (i.e. as the last function, after `on_grpc_close`).